contrib/gorilla/websocket/example_test.go

```diff
@@ -0,0 +1,47 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+package websocket_test
+
+import (
+	"log"
+	"net/http"
+
+	"github.com/gorilla/websocket"
+	muxtrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/gorilla/mux"
+	websockettrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/gorilla/websocket"
+)
+
+// This example illustrates the usage of WrapConn to trace the websocket
+// connections.
+func ExampleWrapConn() {
+	mux := muxtrace.NewRouter()
+
+	var upgrader websocket.Upgrader
+	mux.HandleFunc("/ws", func(w http.ResponseWriter, r *http.Request) {
+		oconn, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			log.Printf("Error on upgrade: %v\n", err)
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+		defer oconn.Close()
+		conn := websockettrace.WrapConn(r.Context(), oconn)
+		for {
+			mt, message, err := conn.ReadMessage()
+			if err != nil {
+				log.Printf("Error reading message: %v\n", err)
+				continue
+			}
+			log.Printf("Received message: %s\n", message)
+
+			if err := conn.WriteMessage(mt, message); err != nil {
+				log.Printf("Error writing message: %v\n", err)
+				continue
+			}
+		}
```

Comment:
Is this a realistic use case?

If I was instrumenting a websocket application, I'd probably think of each `ReadMessage()` as a request that is answered by a follow-up `WriteMessage()`, like shown below:

```go
func wsHandler(w http.ResponseWriter, r *http.Request) {
	c, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Print("upgrade:", err)
		return
	}
	defer c.Close()

	for {
		txSpan := tracer.StartSpan("tx")
		func() {
			defer txSpan.Finish()

			readSpan := tracer.StartSpan("read", tracer.ChildOf(txSpan.Context()))
			mt, message, err := c.ReadMessage()
			if err != nil {
				readSpan.Finish(tracer.WithError(err))
				return
			}
			readSpan.Finish()

			processingSpan := tracer.StartSpan("processing", tracer.ChildOf(txSpan.Context()))
			time.Sleep(500 * time.Millisecond)
			processingSpan.Finish()

			writeSpan := tracer.StartSpan("write", tracer.ChildOf(txSpan.Context()))
			err = c.WriteMessage(mt, message)
			if err != nil {
				writeSpan.Finish(tracer.WithError(err))
				return
			}
			writeSpan.Finish()
		}()
	}
}
``` 

So IMO the API should somehow make this use case easier.

Wrapping the connection to track Read/Write durations by itself seems pretty uninformative, because it doesn't let us associate server-side processing time with the individual messages. Maybe a better API could look like this:

```go
func wsHandler2(w http.ResponseWriter, r *http.Request) {
	c, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Print("upgrade:", err)
		return
	}
	defer c.Close()

	for {
		tx := ddwebsocket.StartTx(c)
		func() {
			defer tx.Finish()

			mt, message, err := tx.ReadMessage() // automatically creates span
			if err != nil {
				return
			}

			processingSpan := tx.StartSpan("processing")
			time.Sleep(500 * time.Millisecond)
			processingSpan.Finish()

			err = tx.WriteMessage(mt, message) // // automatically creates span
			if err != nil {
				return
			}
		}()
	}
}
```

Anyway, I realize that this PR is perhaps more of an exercise than it is the solution to a particular problem, but for the sake of making this a realistic review, it'd be nice to clarify the use case for this so we can figure out if the API is addressing it well : ).

Reply:
Good points indeed. The way the Go context is handled once the Conn created doesn't help (not exposed anymore, and maybe not even used). But I think you can actually do it by doing:

```golang
func wsHandler3(w http.ResponseWriter, r *http.Request) {
	oconn, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Print("upgrade:", err)
		return
	}
	defer oconn.Close()

	// Create a tx span
	tx, ctx := tracer.StartFromContext(r.Context(), "tx")
	defer tx.Finish()
	// Trace the websocket connection under this tx span
	conn := websockettrace.WrapConn(ctx, oconn)
	for {
		mt, message, err := conn.ReadMessage() // automatically creates span
		if err != nil {
			return
		}

		processingSpan := tracer.StartSpanFromContext(ctx, "processing")
		time.Sleep(500 * time.Millisecond)
		processingSpan.Finish()

		err = conn.WriteMessage(mt, message) // // automatically creates span
		if err != nil {
			return
		}
	}
}
```

wdyt?

Reply:
> wdyt?

I was considering that route, but in your example you got a `tx` that has the same lifetime as the `oconn` connection itself. That might be fixable by moving both `WrapConn` as well as `StartFromContext` into the loop, but it seems weird to wrap the same connection multiple times inside of a loop. I mean it could probably work, but it seems really awkward from an API perspective. Do you agree?

Reply:
I don't have enough background on integrations' "api style" to know if we sometimes provides such tracing helpers.
I'll let @gbbr tell us more about that.

Reply:
Sounds good. But just to confirm: You agree that the use case I outlined is reasonable and enabling it should be considered the key problem that this PR is trying to solve?

Reply:
> Sounds good. But just to confirm: You agree that the use case I outlined is reasonable and enabling it should be considered the key problem that this PR is trying to solve?

Actually, the top-level websocket span (tx in our example) is going to be the same as the HTTP request span. So I am not so sure this one is useful. But the inner operation is for sure, and we definitely support this one.

Reply:
> Actually, the top-level websocket span (tx in our example) is going to be the same as the HTTP request span. 

Hm, but that could be a very long span with an unbound number of child spans. Didn't we discuss that this kind of thing doesn't work well with the tracer/agent as of today?

Reply:
Yes, it would be much better if we could separate individual requests/responses over the socket into individual traces. Having the whole connection's activity grouped into one trace is going to be mostly useless for any use case websockets is designed for.

Reply:
Is this possible today?
I felt like the current tracing API misses the ability to "fork" a trace this such case, which would allow to have individual tracers for each sub-request.